packages/gravity/kube-testing/src/spec/replication.ts

```diff
@@ -0,0 +1,412 @@
+          }),
+        );
+      }
+    };
+
+    await testRun();
+
+    // scheduleTaskInterval(
+    //   ctx,
+    //   async () => {
+    //     await env.syncBarrier(`iteration-${testCounter}`);
+    //     await asyncTimeout(testRun(), spec.duration);
+    //     testCounter++;
+    //   },
+    //   spec.repeatInterval,
+    // );
+    // await sleep(spec.duration);
+    // await ctx.dispose();
+
+    log.info('test completed', { agentIdx });
+  }
+
+  async finish(params: TestParams<ReplicationTestSpec>, results: PlanResults): Promise<any> {
+    await this.signalBuilder.destroy();
+    log.info('finished shutdown');
+
+    // Map<agentIdx, Map<swarmId, FeedStats>>
+    const feeds = new Map<number, Map<number, FeedEntry[]>>();
+
+    range(params.spec.agents).forEach((agentIdx) => {
+      feeds.set(agentIdx, new Map(range(params.spec.swarmsPerAgent).map((idx) => [idx, []])));
+    });
+
+    const reader = getReader(results);
+
+    reader.forEach((entry: SerializedLogEntry<any>) => {
+      switch (entry.message) {
+        case 'dxos.test.feed-stats': {
+          const feedStats = entry.context as FeedEntry;
+          const agentFeeds = feeds.get(entry.context.agentIdx)!;
+          const swarmFeeds = agentFeeds.get(entry.context.swarmIdx)!;
+          // TODO(egorgripasov): For some reason received logs entry are duplicated.
+          if (!swarmFeeds.find((feed) => feed.feedKey === feedStats.feedKey)) {
+            swarmFeeds.push(feedStats);
+          }
+          break;
+        }
+      }
+    });
+
+    const stats: FeedStats[] = [];
+    feeds.forEach((agentFeeds, agentIdx) => {
+      agentFeeds.forEach((swarmFeeds, swarmIdx) => {
+        swarmFeeds.forEach((feedStats) => {
+          if (feedStats.writable) {
+            const replicas: FeedReplicaStats[] = [];
+            // Check feeds from other agents from this swarm.
+            feeds.forEach((otherAgentFeeds, otherAgentIdx) => {
+              if (otherAgentIdx !== agentIdx) {
+                const otherSwarmFeeds = otherAgentFeeds.get(swarmIdx)!;
+                const otherFeedStats = otherSwarmFeeds.find(
+                  (otherFeedStats) => otherFeedStats.feedKey === feedStats.feedKey,
+                );
+                if (otherFeedStats) {
+                  const bytesInSecond = otherFeedStats.byteLength / (params.spec.feedLoadDuration / 1000);
+                  const e = Math.floor(Math.log(bytesInSecond) / Math.log(1024));
+                  replicas.push({
+                    feedLength: otherFeedStats.feedLength,
+                    byteLength: otherFeedStats.byteLength,
+                    replicationSpeed:
+                      (bytesInSecond / Math.pow(1024, e)).toFixed(2) + ' ' + ' kMGTP'.charAt(e) + 'B/sec',
+                  });
+                }
+              }
+            });
+            stats.push({
+              feedKey: feedStats.feedKey,
+              feedLength: feedStats.feedLength,
```

Comment:
is feedLength the total length of the feed or the number of mutations replicated?

Reply:
that's the num of mutation, while `byteLength` is the total length in bytes